rosbag2_compression/src/rosbag2_compression/sequential_compression_reader.cpp

```
@@ -0,0 +1,198 @@
+    return;
+  }
+
+  // Currently a bag file can only be played if all topics have the same serialization format.
+  check_topics_serialization_formats(topics);
+  check_converter_serialization_format(
+    converter_options.output_serialization_format,
+    topics[0].topic_metadata.serialization_format);
+}
+
+bool SequentialCompressionReader::has_next()
+{
+  if (storage_) {
+    // If there's no new message, check if there's at least another file to read and update storage
+    // to read from there. Otherwise, check if there's another message.
+    if (!storage_->has_next() && has_next_file()) {
+      load_next_file();
+      storage_ = storage_factory_->open_read_only(
+        *current_file_iterator_, metadata_.storage_identifier);
+    }
+    return storage_->has_next();
+  }
+  throw std::runtime_error{"Bag is not open. Call open() before reading."};
+}
+
+std::shared_ptr<rosbag2_storage::SerializedBagMessage> SequentialCompressionReader::read_next()
+{
+  if (storage_) {
+    auto message = storage_->read_next();
+    if (decompressor_ && compression_mode_ == rosbag2_compression::CompressionMode::MESSAGE) {
+      decompressor_->decompress_serialized_bag_message(message.get());
+    }
+    return converter_ ? converter_->convert(message) : message;
+  }
+  throw std::runtime_error{"Bag is not open. Call open() before reading."};
+}
+
+std::vector<rosbag2_storage::TopicMetadata> SequentialCompressionReader::get_all_topics_and_types()
+{
+  if (storage_) {
+    return storage_->get_all_topics_and_types();
+  }
+  throw std::runtime_error{"Bag is not open. Call open() before reading."};
+}
+
+bool SequentialCompressionReader::has_next_file() const
+{
+  return current_file_iterator_ + 1 != file_paths_.end();
+}
+
+void SequentialCompressionReader::load_next_file()
+{
+  assert(current_file_iterator_ != file_paths_.end());
+  current_file_iterator_++;
+  if (decompressor_ && compression_mode_ == rosbag2_compression::CompressionMode::FILE) {
```

Can decompressor be null in normal operation? If it should never be null, we should not check for that and let the node crash instead of silently losing data.

---

No, it will always throw in `setup_decompression` so yes this is a redundant check.

---

I'm making a change to `setup_compression()` to eagerly throw in the one case where `decompressor_` is not set since it doesn't make sense for `SequentialCompressonReader` to be used on uncompressed bagfiles. I'm then replacing the `nullptr` checks on `decompressor_` with asserts